demo/benchmark.cpp

```diff
@@ -0,0 +1,140 @@
+#include "../src/LockPrimitives.hpp"
+
+#include <iostream>
+#include <atomic>
+#include <mutex>
+#include <future>
+#include <chrono>
+#include <vector>
+
+namespace lab::utils
+{
+    template<typename T>
+    class Incrementable {
+    public:
+        explicit Incrementable(T value) noexcept :
+            _value{std::move(value)}
+        {}
+
+        void increment() noexcept
+        {
+            ++_value;
+        }
+
+        auto value() -> T&
+        {
+            return _value;
+        }
+
+    private:
+        T _value;
+    };
+
+    template<typename T>
+    class Incrementable<std::atomic<T>> {
+    public:
+        explicit Incrementable(T value) :
+                _value{std::move(value)}
+        { }
+
+        void increment()
+        {
+            _value.fetch_add(1, std::memory_order::memory_order_relaxed);
+        }
+
+        auto value() -> std::atomic<T>&
+        {
+            return _value;
+        }
+
+    private:
+        std::atomic<T> _value;
+    };
+
+    template<typename P, typename V>
+    class LockableValue {
+    public:
+        class Proxy {
+        public:
+            explicit Proxy (P& primitive, V& value) :
+                    _lock{primitive},
+                    _value{value}
+            { }
+
+            [[nodiscard]]
+            auto value () -> V&
+            {
+                return _value;
+            }
+
+        private:
+            std::lock_guard<P> _lock;
+            V& _value;
+        };
+
+        template<typename... Args>
+        explicit LockableValue(Args&&...args) :
+                _value{std::forward<Args...>(args...)}
+        { }
+
+        [[nodiscard]]
+        auto lock() -> Proxy
+        {
+            return Proxy{_primitive, _value};
+        }
+
+    private:
+        V _value;
+        P _primitive;
+    };
+
+    constexpr std::size_t N = 1e7;
+    constexpr std::size_t THREADS_NO = 8;
+    template<typename D, typename P, typename V>
+    int benchmark(
+            LockableValue<P, V> lockable_value,
```

Comment:
по значенню тут так собі передавати, хоча б через те, що примітиви синхронізації зазвичай non-copiable
в момент виклику просто копії не було, на місці сконструювалось

Reply:
додам референс

Reply:
Та нормас тут по значенню передавати, нафіга ще десь ззовні створювати окрему змінну, яку сюди по рефу передавати. Ззовні вона, по факту, юзлес. Створювати в тілі функції також шляпа, оскільки caller в цьому випадку бвде це буде записувати як `benchmark<blah-blah>(init-values..);` instead of `benchmark (Lockable<blah-blah>{init-values...});`

Reply:
хай rvalue приймає і все, буде те саме що зараз мінус тупі помилки компіляції коли там і правда копіювання буде